### ISSUE TITLE
How to handle ViewModels with dynamic parameters alongside DI parameters?

### ISSUE THREAD

**Issue by kibiz0r** [simpleinjector/SimpleInjector#643]
I read the previous discussions #58 and #60, but I think my concern is *slightly* (maybe *just* slightly) different and I'm not satisfied with the suggestions from those issues.

Suppose I have a situation like this:

I'm in a Xamarin app, and I want to show a page like "Register a product with model number XYZ and serial number 123". The model number and serial number have already been determined before showing this page, and in order to register the product we need to hit an HTTP endpoint.

So I come up with something like this:

```
public class RegisterProductViewModel
{
    public RegisterProductViewModel(string modelNumber, string serialNumber, IApiClient apiClient)
    {
        this.RegisterCommand = ReactiveCommand.Create(() => apiClient.RegisterProduct(modelNumber, serialNumber));
    }
}
```

That seems to pretty clearly declare the situation:

- The VM's existence is scoped by the modelNumber and serialNumber -- if you have a different combination of them, or if you want to show multiple "Register Product" pages simultaneously, you have to create more VMs.
- In a web context, this is analogous to having a route like `/register-product?modelNumber=XYZ&serialNumber=123`; the last thing we'd wanna do is store a "current model number and serial number" in some global server-side application state, enforcing a modal workflow, so injecting some `IAppState` seems like a bad idea.
- And the VM depends on having an `IApiClient` in order to do its job. So we *do* have to do some dependency injection here.

One possible solution is to have an "Initialize" method, where we pass in the dynamic arguments `modelNumber` and `serialNumber`, and keep the constructor DI-compatible. I don't like that, for a few reasons:

1. That means I have to add a bunch of boilerplate to store dependencies in `private readonly` fields so I can reference them later in the "Initialize" method, where previously I could just do all of my command/property setup in the constructor in a very ["objects are merely a poor man's closures"](http://wiki.c2.com/?ClosuresAndObjectsAreEquivalent) way.
1. It's possible to call that "Initialize" method 0, 1, or N times, and I'd like to guarantee that it only gets called once, as the first method, before anything else is called... which is kind of the definition of a constructor.

Another possible solution is to move the dynamic portion out to the View level, and pass those arguments into the register command on-demand. I don't like that either, because then I'm moving more of the logic away from the model layer and into the far-less-testable realm of UI code. That's reasonable for this case, but it's a slippery slope if I adopt that pattern for the whole codebase.

And like I said earlier, I could inject some global `IAppState`, but that feels gross in a lot of ways. Suddenly, if you're gonna present this ViewModel, you're expected to reach out and manipulate this other totally separate object that you just assume the ViewModel depends on... it's a really hidden indirect contract compared to how obvious it would be if you were required to pass arguments into a constructor.

So for right now, I'm going to just plan on doing this: `new RegisterProductViewModel(modelNumber, serialNumber, container.GetInstance<IApiClient>())`

...but it just doesn't feel right to me that I suddenly have a one-off use of `new` where every other ViewModel is using `GetInstance`.

**Comment by dotnetjunkie** [comment]
To me, your situation isn't different as in #58 or #60 and my answer is the same: prevent mixing runtime data with service dependencies in your components' constructors. My advice is to use an Initialize method on your view model so dependencies and runtime data can be separated.

**Comment by kibiz0r** [comment]
Suppose I don't like that answer, and wish to mix them anyway. 😜 How would I go about implementing a `GetInstance<T>(this Container container, params object[] args)`?

**Comment by dotnetjunkie** [comment]
> Suppose I don't like that answer, and wish to mix them anyway.

What would you like me to say? Use a different container? :)

The answers #58 and #60 already linked to some alternative approaches, but again, prevent yourself the pain and suffering; you chose Simple Injector for a reason, so let it help you. For instance by verifying your configuration. Once you start mixing runtime data and dependencies into constructors, you are on your own and Simple Injector can't verify your graphs anymore.

**Comment by kibiz0r** [comment]
> What would you like me to say? Use a different container? :)

Not at all. I really like the way you've implemented Simple Injector, and I agree with the philosophy you promote. And I appreciate that you include it in the docs. It's so much easier to reference those to coworkers as "the way the tool works" than to have to mount the argument that "Well, Uncle Bob says..." every time.

*What follows is mostly just me formulating my own thoughts, and captured here in case you or someone else just happens to be interested. Feel free to close the issue.*

I guess I'm just not convinced that it's possible to truly divorce runtime data from services. At some point, you need an object that knows about both. You can cheat and retain DI-compatibility by having Foo stash some arguments in a repository-style service and then having Bar retrieve them... but I don't like having those kinds of implicit contracts.

In the past (in Ruby), I've used nested DI contexts, where each level adds an extra set of registrations that are available to that context and any future child contexts. In this way, you still have a directed acyclic graph of dependencies which are all immutable within their context, you avoid looking outside of your class for any sort of state or services, and all the other goodies that come with having DI.

I understand that, being in C#, we have a lot of static information we can inspect, which makes the technical situation pretty different from Ruby. You can trace a dependency graph pretty well using reflection on types and members, but if you start using a `Dictionary<string, object>` to resolve some things then all bets are off.

If we just talk about design, though, and ignore the static analysis implications, what I'm trying to do is very similar to the way SI handles generics.

Consider this somewhat silly but illustrative code:

```
public interface IOrderId
{
    string OrderId { get; }
}

public class OrderId123 : IOrderId
{
    public string OrderId => "123";
}

public interface ICustomerId
{
    string CustomerId { get; }
}

public class CustomerIdXYZ : ICustomerId
{
    public string CustomerId => "XYZ";
}

public class CustomerOrderViewModel<TCustomerId, TOrderId> where TCustomerId : ICustomerId where TOrderId : IOrderId
{
    public CustomerOrderViewModel(TCustomerId customerId, TOrderId orderId, ICustomerOrderItemCache<TCustomerId, TOrderId> itemCache)
    {
    }
}

container.Register(typeof(ICustomerOrderItemCache<,>), typeof(CustomerOrderItemCache<,>));
container.Register(typeof(CustomerOrderViewModel<,>));
var vm = container.GetInstance<CustomerOrderViewModel<CustomerIdXYZ, OrderId123>>();
```

Granted, you'd never wanna do something like this, but the point is that a consumer of `CustomerOrderViewModel` can provide arguments at the time of the `GetInstance` call, and that SI strives to support that kind of invocation -- although perhaps not policy-based design. *shudder*

And, as shown by the `itemCache`, it's possible to propagate those "runtime" arguments (yeah, I know any sane person would be doing generic arguments statically; it's an analogy, not an equivalent solution) to injected dependencies and therefore scope them according to the parent context.

(Note that `CustomerOrderViewModel` should not need to know the type parameters of its item cache. It should just take an `ICustomerOrderItemCache`. It should be up to the cache itself to decide what parameters it needs, and that should be invisible to the ViewModel. You can fix that in a one-off fashion in this case, by doing a custom registration that inspects the injection context of the cache, looking for type arguments that match the cache's type arguments, and forwarding them... but I'm trying to come up with something that does that automatically and goes more than one level deep.)

I think that if I decide to pursue this then I'm gonna have to build a parallel system, probably using an `Initialize` method as you suggested, and try to emulate [Ember's routing system](https://guides.emberjs.com/release/routing/specifying-a-routes-model/#toc_reusing-route-context). It could still interoperate pretty nicely. But I was hoping to leverage some of SI's guts to help with reflection.

I can do a simple, generics-based approach like this:

```
public static class ContainerInitializable
{
    public static ContainerWith<TArg> With<TArg>(this Container container, TArg arg)
    {
        return new ContainerWith<TArg>(container, arg);
    }

    public static ContainerWith<TArg1, TArg2> With<TArg1, TArg2>(this Container container, TArg1 arg1, TArg2 arg2)
    {
        return new ContainerWith<TArg1, TArg2>(container, arg1, arg2);
    }
}

public class ContainerWith<TArg>
{
    public Container Container { get; }
    public TArg Arg { get; }

    public ContainerWith(Container container, TArg arg)
    {
        Container = container;
        Arg = arg;
    }

    public TService GetInstance<TService>() where TService : class, IContainerInitializable<TArg>
    {
        var service = Container.GetInstance<TService>();
        service.Initialize(Arg);
        return service;
    }
}

public class ContainerWith<TArg1, TArg2>
{
    public Container Container { get; }
    public TArg1 Arg1 { get; }
    public TArg2 Arg2 { get; }

    public ContainerWith(Container container, TArg1 arg1, TArg2 arg2)
    {
        Container = container;
        Arg1 = arg1;
        Arg2 = arg2;
    }

    public TService GetInstance<TService>() where TService : class, IContainerInitializable<TArg1, TArg2>
    {
        var service = Container.GetInstance<TService>();
        service.Initialize(Arg1, Arg2);
        return service;
    }
}
```

I'm not sure how far that's really gonna go though, or if it's even right to overlap the two concerns if I'm gonna move towards a router-centric design.

And I probably should've clarified earlier that I don't want to use this kind of "sub-context" mechanism deep inside the model layer of my codebase. Really, I only want it at the routing level, although I guess I could also see an argument for using it in long-lived sub-processes that effectively operate as their own environment.

At any rate, I won't take any more of your time. Unless you like talking about this stuff. Thanks for rubber-ducking for me. 😄
